ortools/sat/timetable_horizontal_edgefinding.cc

```diff
@@ -0,0 +1,290 @@
+        event.time <= window_end) {
+      continue;
+    }
+    window_end = event.time;
+
+    const IntegerValue capacity = integer_trail_->UpperBound(capacity_);
+    if (window_end < ScheduleTasks(window_end, capacity)) {
+      AddScheduleTaskReason(window_end);
+      return helper_->ReportConflict();
+    }
+  }
+
+  return true;
+}
+
+// ScheduleTasks returns a lower bound of the earliest time at which a group 
+// of tasks will complete. The group of tasks is all the tasks finishing before 
+// the end of the window + the fixed part of the tasks having a mandatory part 
+// that overlaps with the window. 
+IntegerValue HorizontallyElasticOverloadChecker::ScheduleTasks(
+    IntegerValue window_end, IntegerValue capacity) {
+  // TODO: If we apply this only by increasing window_end, then there is no
+  // need to re-process the FULL tasks. Only the fixed-part and ignored might
+  // change type. Specifically, FIXED-PART can become FULL while IGNORE can
+  // either become FIXED-PART or FULL. Said otherwise, FULL is the terminal
+  // state.
+  for (int i = 0; i < num_tasks_; ++i) {
+    if (helper_->EndMax(i) <= window_end) {
+      task_types_[i] = TaskType::kFull;
+      continue;
+    }
+    // If the task is external but has a compulsory part that starts before
+    // window_end, then we can process it partially.
+    if (helper_->StartMax(i) < window_end &&
+        helper_->StartMax(i) < helper_->EndMin(i)) {
+      task_types_[i] = TaskType::kFixedPart;
+      continue;
+    }
+    // Otherwise, simply mark the task to be ignored during sweep.
+    task_types_[i] = TaskType::kIgnore;
+  }
+
+  int next_event = 0;
+  IntegerValue time = profile_events_[0].time;
+
+  // Estimation of the earliest time at which all the processed tasks can be
+  // scheduled.
+  IntegerValue new_window_end = kMinIntegerValue;
+
+  // Overload represents the accumulated quantity of energy that could not be
+  // consumed before time.
+  IntegerValue overload = 0;
+
+  // Total demand required at time if all processed tasks were starting at
+  // their start min.
+  IntegerValue demand_req = 0;
+
+  // Total demand required at time if all processed tasks that could overlap
+  // time were. This is used to avoid placing overload in places were no task
+  // would actually be.
+  IntegerValue demand_max = 0;
+
+  while (time < window_end) {
+    // Aggregate the changes of all events happening at time. How to process
+    // an event depends on its type ("full" vs "fixed-part").
+    IntegerValue delta_max = 0;
+    IntegerValue delta_req = 0;
+    while (profile_events_[next_event].time == time) {
+      const ProfileEvent event = profile_events_[next_event];
+      switch (task_types_[event.task_id]) {
+        case TaskType::kIgnore:
+          break;  // drop the event
+        case TaskType::kFull:
+          switch (event.event_type) {
+            case ProfileEventType::kStartMin:
+              delta_max += event.height;
```

Comment:
nit: I think changing demand_req and demand_max directly without the deltas might be simpler to understand.

Reply:
Agreed, done 👍